### DISCUSSION TITLE
Not working on Edge

### DISCUSSION THREAD

**Discussion by jdgamble555** [etherCorps/sveltekit-og#41] [General]
I'm not getting this to work on Vercel Edge.  Is anyone else experiencing this?  Is this a Vercel problem?

https://github.com/orgs/vercel/discussions/5250

Getting this font `noto-sans-v27-latin-regular.ttf` error.

J

**Comment by theetherGit** [comment]
I have already  tried to make it work with edge but i have failed to do so.

**Reply by jdgamble555** [reply]
What about Cloudflare?

**Reply by theetherGit** [reply]
Cloudflare workers are also part of edge services.

**Reply by jdgamble555** [reply]
Right. Just wondering if different results.  Hope they fix this :)

**Reply by jdgamble555** [reply]
It looks like they may have fixed cloudflare - https://github.com/cloudflare/pages-plugins/issues/36#event-12020090940

**Reply by theetherGit** [reply]
I have to test it and I'm not free before weekend.

**Reply by RickRyan26** [reply]
Any forward progress on getting sveltekit-og working on the edge? I would LOVE for this to happen. Feels like Vercel is intentionally holding Svelte back on this one...

**Reply by theetherGit** [reply]
Not yet, even though I want this to work on every runtime but I can't help. I have tried so many things on this but it didn't work out. Currently i want someone to take a look at this, as I'm not able to think of a solution on this.

If anyone wants to help with this please let me know. We can brainstorm a bit on this.

**Comment by jdgamble555** [comment]
Have you tried building off of satori? Looks like Nuxt did this - https://nuxt.com/modules/og-image

**Reply by theetherGit** [reply]
Oh, no i haven't. Just gone through there implementation a bit different approach but I will try with test deployment on cloudflare workers if works then I'll start to get things from here.

**Reply by RickRyan26** [reply]
FYI Lee Rob posted on X a few days ago saying they're moving Vercel off the Edge runtime (Cloudflare under the hood) because Node 20 is somehow faster than it. The Edge runtime is also different from Vercel's CDN which they confusingly also call the Edge. Correct me if I'm wrong but I now believe svelteKit-og should intentionally avoid the Edge runtime yet include guidelines on how to easily cache the images it creates on a CDN using Vercel's _vercel/ urls or whatever.

**Reply by theetherGit** [reply]
Yes, the vercel edge now uses v20, still cloudflare workers is something i use day to day and i wanna support that (i feel that workerd going to live long) and for caching I'll have to see.

**Reply by RickRyan26** [reply]
Wait... Vercel Edge doesn't use v20, Vercel's website does, right?

**Reply by theetherGit** [reply]
https://twitter.com/leeerob/status/1780705942734331983?t=T7eoKiC2Ydo2xULLII-WXw&s=19

**Reply by jdgamble555** [reply]
Vercel wants to use a combination of edge and serverless. Serverless supports node v18 or v20, while edge is just cloudflare workers. I personally moved from vercel to cloudflare due to pricing. Serverless has cold starts which will always be a problem. Either way, edge functions (cloudflare workers or deno in some environments) need a way to support og/image. For more on Vercel - https://youtu.be/lAGE-k1Zfrg?si=bQWFnhzN8-OmjASw

**Reply by theetherGit** [reply]
I read that that's what they mean by edge compute and rendering they shifted this to node v18 and v20 and edge network is just a cdn for serving some html css js files.

**Reply by RickRyan26** [reply]
@theetherGit I think you're mistaken, switching the runtime on people would cause spontaneous bugs. Vercel themselves has just stopped using it internally which implies that the Edge is no longer the bleeding Edge.

@jdgamble555 I watched that video and in the tweet it says "We tested it extensively with @v0. Using the Node.js runtime with 1 vCPU (our "Standard" performance) option for Vercel Functions had **faster startups** 🤯". I also heard somewhere that AWS has a solution to the cold start problem.

**Reply by jdgamble555** [reply]
faster startups than node 18, edge computing has no startups

**Reply by theetherGit** [reply]
> I think you're mistaken, switching the runtime on people would cause spontaneous bugs. Vercel themselves has just stopped using it internally which implies that the Edge is no longer the bleeding Edge.
Yes, I got it wrong.

But why are we going off topic here, we still needs to implement `sveltekit-og` for `edge runtimes`. Let's just focus on how we can implement it rather than this. We should come with good ideas till friday and I'll work on these on this saturday. Let's make it work, that's the target for this weekend if you can help with implementation, suggestion for caching anything it will be good for this project.

**Reply by RickRyan26** [reply]
@jdgamble555 I read it a few times now and I really think He's comparing Node in general with Edge, ChatGPT agrees. The reason for the Tweet is because we basically all believed the edge was the faster option.

@theetherGit Vercel has good documentation for their Image Optimization. Just need to add some params to the vercel.json config and use your final og URLs as a query inside a special _vercel URL. Didn't mean to derail the thread, just saying what SvelteKit-og already has is the fastest option.

**Reply by theetherGit** [reply]
@RickRyan26 
> need to add some params to the vercel.json config and use your final og URLs as a query inside a special _vercel URL

This needs to be in documentation, so whoever is deploying to vercel will configure that as per there requirements. And if you can provide the documentation URL here will be better I'll go through that.

I can provide helper function that can make caching better for e.g. 
- For cloudflare workers I can provide a function that takes event and cloudflare cache object as args and that function will add that to cloudflare cache and other function that checks for cache in Cloudflare Cache Object.

**Reply by RickRyan26** [reply]
Here's the documentation that shows the URL and query params to use.
https://vercel.com/docs/image-optimization

Note they don't have a SvelteKit version so you have to go with the build your own framework route and add the following image config to the Vercel JSON file as seen here.
https://vercel.com/docs/projects/project-configuration#images

It's much easier than the documentation makes it seem.

**Reply by jdgamble555** [reply]
That is image optimization, not og/image

**Reply by RickRyan26** [reply]
It puts the og/image on vercels edge network, aka cdn

**Reply by jdgamble555** [reply]
This does not solve the problem. BTW Vercel Image optimization is extremely expensive, and not even necessary here. When you add cache headers, things automatically get added to the CDN. That is not the problem here. The problem is being able to generate the images if you host on Cloudflare or don't use Serverless.

**Reply by RickRyan26** [reply]
You're blowing my mind rn 😂 So you're saying the below will save the og image's to the CDN in theory?

```
// src/hooks.js
export async function handle({ request, resolve }) {
 const response = await resolve(request);
 if (request.path.startsWith('/og/')) {
    response.headers['Cache-Control'] = 'public, max-age=31536000, immutable';
 }
 return response;
}
```

Marketing language is going to be the end of me!

**Reply by jdgamble555** [reply]
I believe so yes, although I'm not going to judge you for not using TS. You have to return an image type, as if the file were an image I believe. Something like:
```ts
const data = await response.blob();

return new Response(data, {
    headers: {
        'content-type': 'image/png',
        'cache-control': 'public, immutable, no-transform, max-age=31536000'
    }
});
```

**Reply by theetherGit** [reply]
@RickRyan26 I got it how vercel cdn works but i don't need to do anything for that. As vercel.json config needs to be handled by developers. So we just need to mention in docs with examples how to do it. I don't need to implement anything in Library itself.

@jdgamble555 your provided code is not wrong but it is mainly meant for browsers not servers so only visited users will have that image in their own local cache not on the server. But in cloudflare scenario you can utilise same headers to cache images on there cache network.

For cloudflare developers can use Cache object cache an endpoint on cloudflare network.

From my PoV vercel image cdn somewhat similar to cloudflare Images but overpriced.

Other than that I don't think there is anything related to cache that i need to worry about.

**Reply by jdgamble555** [reply]
That code is meant to be in a `+page.server.ts` or `+server.ts` file. The caching depends on the headers. There are specific headers for cdns like `Cloudflare-CDN-Cache-Control`, but generally regular caching can handle more than one level. Look at the specifics to your cdn. Vercel Edge is Cloudflare btw (workers and pages) while vercel serverless is aws lambdas. I don't use serverless at all, so that is my problem at hand.

**Comment by theetherGit** [comment]
@jdgamble555 Thanks for Nuxt did this - [nuxt.com/modules/og-image](https://nuxt.com/modules/og-image) - I got pretty good idea how to handle things internally.

Any of you wanna help with designing docs for sveltekit-og and svelte-h2j?

**Reply by DarthGigi** [reply]
Hey, I'd like to help out with the documentation website

**Reply by theetherGit** [reply]
@DarthGigi You have any ui example? or design in mind?

**Reply by DarthGigi** [reply]
@theetherGit I think using https://starlight.astro.build would be a good start. It's build by the makers of Astro and Astro itself supports writhing .svelte files/code if the projects needs it.
And the actual documentation content is written in markdown/mdx.

**Reply by theetherGit** [reply]
@DarthGigi I wanted to but i wanna stick to sveltekit only IK it adds more work and workarounds to do things but I like
https://formsnap.dev/ by https://github.com/svecosystem And i want something similar. I did ping [huntabyte](https://github.com/huntabyte) once and he said he will try to make a template or something but didn't heard anything after that. You can try to connect with him over X.

**Reply by DarthGigi** [reply]
@theetherGit I see. I guess we could also just clone the docs folder and change the things we want? https://github.com/svecosystem/formsnap/tree/main/sites/docs

**Reply by theetherGit** [reply]
@DarthGigi I planned to do that but didn't had that much time to go through all things in the repo. If you can and provide a demo, then we can make it a template for docs.

**Reply by DarthGigi** [reply]
@theetherGit Sure, I'll see what I can do. I don't have a lot of time on my hands these days, so it might take a while before I can even get started on it 😅

**Reply by theetherGit** [reply]
Ohh, Let's see. I'm not free for next three weekend. That's the main problem (I'm just trying to fix og in weekdays. From 12 May to 31 May, I'm on a break.)

**Reply by jdgamble555** [reply]
@theetherGit - It looks like you rewrote the module to use just `satori`, were you able to see what the nuxt module does in order to get it working on edge functions?

**Reply by jdgamble555** [reply]
@theetherGit - I take it you weren't... well, anyway, here is a working plugin in case you want to update your package - https://github.com/fineshopdesign/cf-wasm/issues/60

**Reply by theetherGit** [reply]
Ok, I'll give it a try.

**Reply by theetherGit** [reply]
@jdgamble555 https://github.com/etherCorps/sveltekit-og/issues/47#issuecomment-3557259389
